Improve handling of user flags in configure

vadz opened the pull request:
@ojwb I wonder if you might have any comments about this?

Only the changes to `configure.in` and `build/bakefiles/*.bkl` are interesting here, the rest is auto-generated from them.

vadz committed:
Get rid of WXCONFIG_ONLY_CPPFLAGS variable in configure

This variable was set and used exactly once, so just remove it to
simplify things.

No real changes.

vadz committed:
Don't override CFLAGS etc in configure-generated makefile

CPPFLAGS, CFLAGS, CXXFLAGS and LDFLAGS are supposed to be under
user-control and putting configure-determined options in them broke
something as simple as running "make CXXFLAGS=-Wno-some-extra-warning"
because this overrode the CXXFLAGS set by configure and required for
build.

Improve this by using WX_*FLAGS in the generated makefile and leaving
the user-controlled FLAGS alone. This is still not ideal as running
"configure CFLAGS=-DFOO" and then "make CFLAGS=-DBAR" will define both
FOO and BAR, as configure copies CFLAGS to WX_CFLAGS, and so setting it
on make command line won't override it, as it should, but this should be
a much more rare and also much less severe problem, so we should be able
to live with it for now.

Normally this commit shouldn't result in any user-visible changes, i.e.
it shouldn't break any previously working scenarios and only make some
previously broken ones work.

ojwb commented on configure:
```diff
@@ -37726,8 +37724,9 @@ CXXFLAGS=`echo $WXCONFIG_CXXFLAGS $C_AND_CXX_FLAGS $CXXFLAGS `
-
+if test "SHARED" = 1; then
+    WXCONFIG_CPPFLAGS="$WXCONFIG_CPPFLAGS -DWXUSINGDLL"
+fi
```
I notice this no longer uses the `echo $X` "pattern", which appears to be done to "remove the extra white space" according to a comment elsewhere.  However, you can't safely just `echo $X` since echo itself takes command line arguments on some platforms - e.g. on linux with bash:

```
$ WXCONFIG_CFLAGS='-e -f -g'
$ WXCONFIG_CFLAGS=`echo $WXCONFIG_CFLAGS`
$ export WXCONFIG_CFLAGS
$ printenv WXCONFIG_CFLAGS
-f 
```

There are also issues with some shells if the value contains backslashes.

This change actually fixes such bugs for this case, but more generally it's probably better not to do this at all - extra whitespace is just cosmetic, but losing options like above could be really problematic.  But if the whitespace cleaning is felt to be desirable, something like `printf '%s' $WXCONFIG_CFLAGS` would probably be safe (the autoconf manual suggests that's portable).

vadz replied:
Thanks, this is a good point, and if autoconf thinks `printf` is portable, we should definitely use it.  I'll do it in another commit.

ojwb commented:
> ojwb I wonder if you might have any comments about this?

It does seem an improvement, though as you say in the commit message it's still not quite right.

The debian packaging doesn't override flags at `make`-time, if that's what you're asking - we specify them on the configure command like so:

```
COMMON_CONFIGURE_OPTIONS := \
    --prefix=/usr \
    --with-flavour=$(DEBIAN_WXFLAVOUR) \
    --with-zlib=sys \
    --with-expat=sys \
    $(shell DEB_LDFLAGS_MAINT_APPEND=-Wl,--as-needed DEB_CXXFLAGS_MAINT_APPEND=-std=gnu++98 dpkg-buildflags --export=configure)
```

And that `$shell(...)` expands something like this:

```
$ DEB_LDFLAGS_MAINT_APPEND=-Wl,--as-needed DEB_CXXFLAGS_MAINT_APPEND=-std=gnu++98 dpkg-buildflags --export=configure
CFLAGS="-g -O2 -fdebug-prefix-map=/home/olly=. -fstack-protector-strong -Wformat -Werror=format-security" CPPFLAGS="-Wdate-time -D_FORTIFY_SOURCE=2" CXXFLAGS="-g -O2 -fdebug-prefix-map=/home/olly=. -fstack-protector-strong -Wformat -Werror=format-security -std=gnu++98" FCFLAGS="-g -O2 -fdebug-prefix-map=/home/olly=. -fstack-protector-strong" FFLAGS="-g -O2 -fdebug-prefix-map=/home/olly=. -fstack-protector-strong" GCJFLAGS="-g -O2 -fdebug-prefix-map=/home/olly=. -fstack-protector-strong" LDFLAGS="-Wl,-z,relro -Wl,--as-needed" OBJCFLAGS="-g -O2 -fdebug-prefix-map=/home/olly=. -fstack-protector-strong -Wformat -Werror=format-security" OBJCXXFLAGS="-g -O2 -fdebug-prefix-map=/home/olly=. -fstack-protector-strong -Wformat -Werror=format-security"
```

While looking I noticed a problematic idiom in use in `configure.in` though, which I've commented about.

vadz commented:
> It does seem an improvement, though as you say in the commit message it's still not quite right.

Unfortunately making it 100% right seems to be difficult: the only way to do it I see is to check all the places where `CFLAGS` is changed in configure and either change `WX_CFLAGS` instead or, and in many places we really do need to do this, change both of them, as `CFLAGS` still must be changed to affect further configure tests.

> The debian packaging doesn't override flags at make-time, if that's what you're asking

My question was more general, i.e. I just thought you must have a lot of experience with this stuff, but this at least explains why Debian packages were not affected by this.

Thanks for the review!

ojwb commented:
More generally, I'd say it is a lot more common to specify flags at configure-time than to override them at make-time, and some other packages incorrectly put probed flags into `CFLAGS` or similar - in general I wouldn't be entirely surprised to find overriding at make-time doesn't quite work as it is meant to, and I'd definitely set them at configure-time in preference where possible.

vadz commented:
The trouble is that I do _not_ want to override anything, but just building wx as a dependency from "build everything" makefile doesn't work if you do `make CXXFLAGS=-Wno-some-stupid-warning`. This is annoying...

ojwb commented:
I guess I really meant "specifying them at make-time" there.

vadz commented:
I must be blind, but I don't understand why does this [Travis CI job](https://travis-ci.org/wxWidgets/wxWidgets/jobs/644930433) fail. The link command line seems to be almost identical to the successful build from the current master (this change just moves `-pthread` around), yet it doesn't work.

I can reproduce this locally, too, so it's not just some Travis weirdness: linking minimal sample fails with
```
/usr/bin/ld: /...full path to build dir.../lib/libwx_gtk2u_core-3.1.a(corelib_imagpng.o): undefined reference to symbol 'png_read_end@@PNG16_0'
/usr/bin/ld: //usr/lib/x86_64-linux-gnu/libpng16.so.16: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
```

But `nm --with-symbol-versions` doesn't even show the versioned symbol, and the unversioned one is defined in `libwxpng-3.1.a` which we link with.

Does anybody have any idea about what's going wrong here?

vadz commented:
Oh, as usual, it was enough to ask. I was indeed blind and didn't notice that PNG symbols renaming didn't take place in the new build. Now I just have to find why was this the case.

vadz committed:
Don't use GUI 3rd party libraries for non-GUI libraries

Don't use include paths for jpeg, png and tiff libraries when building
non-GUI projects, such as base, net or xml, and don't make these
libraries depend on the GUI 3rd party libraries neither.

No real changes, just simplify things a bit.

vadz committed:
Ensure that include paths to 3rd party libraries come first

Rearrange wx_{lib,dll} templates definitions to put 3rd party include
paths in the beginning, before WX_CPPFLAGS inherited from wx_{lib,dll}_b
template, so that our own 3rd party libraries headers are found before
the system ones, if we're using built-in versions of the libraries.

vadz commented:
Note to self: this removes `CXXWARNINGS` but it's still used in `tests/test.bkl`, which needs to be updated now.

vadz committed:
Stop using non-existent CXXWARNINGS for failing tests checks

This variable doesn't exist any more, but now it's possible to use
CPPFLAGS directly and get rid of the hack which required it to be used
in the first place.

vadz committed:
Remove duplicated flags from {C,CPP,CXX,LD}FLAGS in configure

This results in much shorter compilation lines and hence much shorter
build logs, which makes examining them simpler.

There should be no real changes, as we should never use duplicated
options intentionally.

MaartenBent commented:
Hi, looks like this brakes gtk3 webview build (and maybe the SDL thing too? it has similar flags):

```
/home/maarten/dev/build/wxWidgets/cfg/bk-deps g++ -c -o webkit2_ext_webview_webkit2_extension.o -DWXUSINGDLL -DwxUSE_GUI=0 -fPIC -DPIC   ../../../wxWidgets/src/gtk/webview_webkit2_extension.cpp
../../../wxWidgets/src/gtk/webview_webkit2_extension.cpp:9:10: fatal error: wx/defs.h: No such file or directory
    9 | #include "wx/defs.h"
      |          ^~~~~~~~~~~
```

vadz commented:
Oops. it turns out I didn't have `libwebkit2gtk-4.0-dev` installed and so didn't use WebKit2 at all when building. But I've installed it now and still don't get any errors because for me `WEBVIEWDLL_CXXFLAGS` contains `$(WX_CPPFLAGS)` and, via it, all the required `-I` options. I'm really not sure why this isn't the case for you. Could you please share you full configure command line? TIA!

MaartenBent commented:
Just `configure --with-gtk=3`. And `WEBVIEWDLL_CXXFLAGS` is fine, but `WEBKIT2_EXT_CXXFLAGS` is not.

AliKet commented:
I got the same error as Maarten!
with `configure --with-gtk=3`

vadz commented:
Sorry, I've finally fixed this in eb7b84f781358574d8e4bc8cd0fac95581422b7f. Thanks for noticing and patiently explaining me what the problem really way, I had completely missed it originally.
